z3c/dependencychecker/db.py

```diff
@@ -0,0 +1,85 @@
+# -*- coding: utf-8 -*-
+import logging
+
+
+logger = logging.getLogger(__name__)
+
+
+class ImportsDatabase(object):
+
+    def __init__(self):
+        self._requirements = []
+        self._extras_requirements = {}
+
+    def set_requirements(self, requirements_list):
+        for requirement in requirements_list:
+            if self._requirement_already_exists(requirement):
+                logger.warn(
+                    '"%s" is declared twice on setup.py',
+                    requirement.name,
+                )
+            else:
+                self._requirements.append(requirement)
+
+    def _requirement_already_exists(self, requirement):
+        names = self.get_requirements_names()
+        return bool(requirement.name in names)
+
+    def get_requirements_names(self):
+        return [
+            x.name
+            for x in self._requirements
+        ]
+
+    def set_extra_requirements(self, extra_name, dotted_names):
+        """A bit of extra work needs to be done as pkg_resources API returns
+        all common requirements as well as the extras when asking for an extra
+        requirements.
+        """
+        only_extra_dotted_names = self._remove_duplicates(dotted_names)
+        if extra_name in self.get_extras():
+            new_dotted_names = self._only_new_requirements(
+                extra_name,
+                dotted_names,
+            )
+            self._extras_requirements[extra_name] += new_dotted_names
+
+            logger.warn(
+                'extra requirement "%s" is declared twice on setup.py',
+                extra_name,
+            )
+        else:
+            self._extras_requirements[extra_name] = only_extra_dotted_names
+
+    def _remove_duplicates(self, imports, other=None):
```

Comment:
What does this do? Removing duplicates can be done via python sets.

The method name is also a bit confusing. "remove duplicates" sounds like it removes duplicates from some existing variable.
What you seem to be doing is more to filter an input variable and return a filtered list without duplicates.
Something like `._cleaned_requirements()` sounds better to me.

Reply:
Similarly: what is `other`? Do you know a clearer name for that?

And if not set, it is the list of the regular requirements' names. But shouldn't 'other' be appended to that regular list? The name doesn't make that clear.

Reply:
I refactored the whole method:
- added a docstring
- renamed it to ``_filter_duplicates``
- rename the argument ``other`` to ``reference``
- actually we can remove the ``get_requirements_names`` as everything is a DottedName, there is no need to be passing around plain strings, as DottedNames already have all the dunder methods needed

Rather than cleaning I decided to use ``filter`` as it is really a matter of getting the items from "imports" that are not found in "other/reference"

Thanks for pointing things out so that I have to have a second look at the code, just with what I reviewed from your comments I could remove at least one or two methods and make the others far simpler! 🙇